Rewrite casts in between predicate

findinpath opened the pull request:
<!-- Provide an overview for maintainers and reviewers. -->
## Description

Rewrite a cast that appears as the value of a `BETWEEN` predicate into a range over the underlying expression, so the predicate can be pushed into the scan. For example, given `t::timestamp(6)`:

    CAST(t AS date) BETWEEN DATE '2022-01-01' AND DATE '2022-01-02'

is rewritten as

    t BETWEEN TIMESTAMP '2022-01-01 00:00:00.000000' AND TIMESTAMP '2022-01-02 23:59:59.999999'

Each half of the `BETWEEN` is unwrapped independently. A strict bound produced by the unwrapping (for example `t < TIMESTAMP '...start of next day...'`) is tightened to its inclusive neighbor using `Type.getPreviousValue`/`getNextValue`, and the two bounds are recombined into an inclusive range. Bounds that fall outside the source type range collapse to a constant truth value: an always-true bound drops out, and an empty or unsatisfiable range becomes `false` for a non-null source.

When a bound cannot be tightened to an inclusive neighbor (a source type such as `real` has no adjacent value), the two comparisons are kept as a conjunction that references the source in both halves. A non-trivial source is bound once via a `Let` so it is evaluated a single time; a column reference stays inline so the bare comparisons remain available for pruning.

A range predicate over the raw column can be turned into a `TupleDomain`, which connectors with min/max metadata (such as Iceberg manifest and data file statistics) use for partition and file pruning. That range representation is not available while the predicate is left as a cast.

## Additional context and related issues

Alternative to #14452, built on the work in #12797 and #14390. The `Between` IR node has since been removed, so the rewrite runs over the `Let` form that `BETWEEN` desugars into (`UnwrapCastInComparison.Visitor#rewriteLet` and `unwrapCastInBetween`); the original author's boundary logic and test cases are preserved.

Rebased on #30286, which fixes the same class of source duplication for plain timestamp-to-date comparisons by binding a non-trivial source once via `Let` instead of duplicating it. This PR reuses that `bindSourceIfNecessary` helper for the `BETWEEN` conjunction fallback described above.

Correctness is covered by `io.trino.sql.planner.TestUnwrapCastInComparison#testBetween`, `#testBetweenDouble`, and `io.trino.sql.query.TestUnwrapCastInComparison`.

<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
## General
* Improve performance of queries with `BETWEEN` predicates that compare a cast expression against a range of literals, such as `CAST(ts AS date) BETWEEN DATE '2020-01-01' AND DATE '2020-01-02'`. ({issue}`14648`)
```

findinpath commented:
**Current limitation**

I stumbled while trying to add the methods:

- `getPreviousValue(Object)`
- `getNextValue(Object)`

to `DateType`.
I am assuming that checking the validity of a date needs joda-time expertise which is not available in `trino-spi` module.

As an alternative we could use custom operators for `nextval`, `previousval` included in `DateOperators` (and the other operators for other types as well) to cover this need.

findinpath commented on core/trino-main/src/test/java/io/trino/sql/planner/TestUnwrapCastInComparison.java:
```diff
@@ -297,6 +297,84 @@ public void testGreaterThanOrEqual()
+        // non-representable, min cast round down, max cast no rounding
+        testUnwrap("bigint", "a BETWEEN DOUBLE '1.1' AND DOUBLE '2'", "a BETWEEN BIGINT '2' AND BIGINT '2'");
+        // non-representable, min cast round up, max cast round down
+        testUnwrap("bigint", "a BETWEEN DOUBLE '1.9' AND DOUBLE '2.2'", "a BETWEEN BIGINT '2' AND BIGINT '2'");
```
This `BETWEEN` predicate can be turned into an equality.
I didn't do it yet because the PR is currently just a draft.

findinpath commented on core/trino-spi/src/main/java/io/trino/spi/type/LongTimestampWithTimeZoneType.java:
```diff
@@ -175,7 +175,7 @@ public Optional<Object> getNextValue(Object value)
                 return Optional.empty();
             }
-            epochMillis--;
+            epochMillis++;
```
Commit spined off to https://github.com/trinodb/trino/pull/18588

martint commented:
I assume this is a typo, but in the motivating example the rewrite should be:

```
 t >= TIMESTAMP '2022-01-01 00:00:00.000000' AND t < TIMESTAMP '2022-01-03 00:00:00.000000'
```

(Note the precision of the first timestamp)

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -172,6 +175,117 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
+                return expression;
+            }
+
+            ComparisonExpression lowBound = new ComparisonExpression(GREATER_THAN_OR_EQUAL, node.getValue(), node.getMin());
+            ComparisonExpression highBound = new ComparisonExpression(LESS_THAN_OR_EQUAL, node.getValue(), node.getMax());
```
Don't synthesize expression for these comparisons. Instead, the underlying methods should be adjusted to not have to take a ComparisonExpression. See https://github.com/trinodb/trino/pull/18138#discussion_r1253337523

findinpath replied:
Reused logic from #18138 

Thank you @kabunchi

findinpath commented:
Rebasing on top of `master` to reuse https://github.com/trinodb/trino/pull/18588

Going to _Draft mode_ until the comments from the PR are addressed.

findinpath commented:
Failure on ci/test-other-modules

```
Error:  io.trino.plugin.google.sheets.TestGoogleSheets.init -- Time elapsed: 1.023 s <<< FAILURE!
com.google.api.client.googleapis.json.GoogleJsonResponseException: 
503 Service Unavailable
POST https://sheets.googleapis.com/v4/spreadsheets?fields=spreadsheetId
{
  "code" : 503,
  "errors" : [ {
    "domain" : "global",
    "message" : "The service is currently unavailable.",
    "reason" : "backendError"
  } ],
  "message" : "The service is currently unavailable.",
  "status" : "UNAVAILABLE"
}
```

findepi commented:
> I assume this is a typo, but in the motivating example the rewrite should be:
> 
> ```
>  t >= TIMESTAMP '2022-01-01 00:00:00.000000' AND t < TIMESTAMP '2022-01-03 00:00:00.000000'
> ```

I think it might be desirable to have `<=` on the right, as this allows keeping `BETWEEN` as the predicate.

mosabua commented:
Is this still in progress @findinpath ?

findinpath commented:
@mosabua I did try repeatedly to find a maintainer to accompany this PR without success.

@martint  can you suggest somebody to accompany eventually this PR ?

martint commented:
@findinpath, sorry, this fell between the cracks. I'm happy to help get it in. Can you rebase?

findinpath commented:
@martint  I did rebase on top of the latest changes.
Please have a look.
Feel free to comment if the issue description is not making clear the purpose of this PR.

Heads'up: As a follow-up, i'll make a similar improvement for `date_trunc` (after reviving https://github.com/trinodb/trino/pull/14451)

findinpath commented:
Unrelated CI failure https://github.com/trinodb/trino/issues/20391

findinpath committed:
Adjust unwrapCast method to not be dependent on ComparisonExpression

Co-authored-by: kabunchi <tal@varada.io>

findinpath committed:
Unwrap casts in BETWEEN predicate

This change allows the engine to infer that, for instance,
given t::timestamp(6)

    cast(t as date) BETWEEN DATE '2022-01-01' AND DATE '2022-01-02'

can be rewritten as

    t BETWEEN TIMESTAMP '2022-01-01 00:00:00' AND TIMESTAMP '2022-01-02 23:59:59.999999'

Range predicate `BetweenPredicate` can be transformed into a `TupleDomain`
and thus help with predicate pushdown.
Range-based `TupleDomain` representation is critical for connectors
which have min/max-based metadata (like Iceberg manifests lists which
play a key role in partition pruning or Iceberg data files), as ranges allow
for intersection tests, something that is hard
to do in a generic manner for `ConnectorExpression`.

findinpath committed:
Unwrap casts in BETWEEN predicate

This change allows the engine to infer that, for instance,
given t::timestamp(6)

    cast(t as date) BETWEEN DATE '2022-01-01' AND DATE '2022-01-02'

can be rewritten as

    t BETWEEN TIMESTAMP '2022-01-01 00:00:00' AND TIMESTAMP '2022-01-02 23:59:59.999999'

Range predicate `BetweenPredicate` can be transformed into a `TupleDomain`
and thus help with predicate pushdown.

Co-authored-by: Raunaq Morarka <raunaqmorarka@gmail.com>

findinpath committed:
empty

findinpath committed:
empty

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+            Expression lowBoundUnwrapped = optionalLowBoundUnwrapped.get();
+            Expression highBoundUnwrapped = optionalHighBoundUnwrapped.get();
+
+            Expression trueIfNotNullExpression = trueIfNotNull(cast.expression());
+            if (trueIfNotNullExpression.equals(lowBoundUnwrapped)) {
+                return highBoundUnwrapped;
+            }
+            if (trueIfNotNullExpression.equals(highBoundUnwrapped)) {
+                return lowBoundUnwrapped;
+            }
```
Not that this is potentially incorrect. Consider
- `lowBoundUnwrapped`: `trueIfNotNull(cast)`
- `highBoundUnwrapped`: evaluates to `true` for null value of `cast`.

I understand this is not the case today because the `lowBoundUnwrapped` and `highBoundUnwrapped` can only take a few different forms, as they are results of the `unwrapCast` method.

However, the `unwrapCast` method might evolve over time, and this logic might get out of sync.

I suggest that before we proceed to combining `lowBoundUnwrapped` and `highBoundUnwrapped`, we first assert that both expressions satisfy some structural requirements.

raunaqmorarka replied:
Addressed by the rewrite. Each bound is now classified independently: `isNeverSatisfied` short-circuits to `falseIfNotNull(source)`, a bound that always holds drops out leaving the opposite bound, and only a comparison of the source against a non-null constant feeds the range rebuild. No assumption that both bounds share a form.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+            Expression falseIfNotNullExpression = falseIfNotNull(cast.expression());
+            if (falseIfNotNullExpression.equals(lowBoundUnwrapped) || falseIfNotNullExpression.equals(highBoundUnwrapped)) {
+                if (falseIfNotNullExpression.equals(lowBoundUnwrapped) && falseIfNotNullExpression.equals(highBoundUnwrapped)) {
+                    return falseIfNotNullExpression;
```
Is each of these cases tested?

raunaqmorarka replied:
Yes. `testBetween` now covers a low or high bound dropping out, both bounds vacuous (true for a non-null source), a never-satisfied bound, and an empty range.

kasiafi reviewed (commented):
Just started reviewing

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+                            return expression;
+                        }
+
+                        int compareLowBoundValueAndHighBoundValue = compare(sourceType, lowBoundValue, highBoundValue);
```
The `compare` method would throw exception if `lowBoundValue` or `highBoundValue` was null. 

Again, we are making implicit assumptions about what the `unwrapCast` method returns.

raunaqmorarka replied:
`comparisonBound` requires a non-null constant, so a null value produces no comparison bound and `compare` is never reached.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+                            if (compareNextLowAndHighBound > 0) {
+                                return falseIfNotNull(cast.expression());
+                            }
+                        }
```
Could we refactor the preceding code for low and high bounds of the created between expression so that we
1. compute the low and high bounds (either as the constants from the comparisons or the next/previous values)
2. compare them and return `falseIfNotNull(cast.expression())` if low > high
3. create and return the between expression

?

raunaqmorarka replied:
Done. `inclusiveLowBound` and `inclusiveHighBound` compute the bounds (the constant, or its next/previous value for a strict comparison), then they are compared, returning `falseIfNotNull(source)` when low > high, otherwise the inclusive BETWEEN is built.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+                    }
+                }
+
+                return and(lowBoundUnwrappedComparison, highBoundUnwrappedComparison);
```
Is each case covered in tests?

raunaqmorarka replied:
Covered by the new `testBetween` cases.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+                return and(lowBoundUnwrappedComparison, highBoundUnwrappedComparison);
+            }
+
+            return expression;
```
The `unwrapCast` method might return constant `FALSE`. I don't think this case is covered. The result would be `FALSE`.

raunaqmorarka replied:
`isNeverSatisfied` now treats both `FALSE` and `falseIfNotNull(source)` as unsatisfiable, collapsing the whole range to `falseIfNotNull(source)`. Covered by the low-bound-above-range case in `testBetween`.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+                Type highBoundType = highBoundUnwrappedComparison.right().type();
+
+                if (cast.expression().equals(lowBoundUnwrappedComparison.left()) &&
+                        Objects.equals(sourceType, lowBoundType) &&
```
Removed. The rewrite no longer compares operand types; it matches the source expression by equality and reads the constant value.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+                if (cast.expression().equals(lowBoundUnwrappedComparison.left()) &&
+                        Objects.equals(sourceType, lowBoundType) &&
+                        cast.expression().equals(highBoundUnwrappedComparison.left()) &&
+                        Objects.equals(sourceType, highBoundType)) {
```
Same, the type comparison is gone.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+                        // Try to reconstruct the BETWEEN predicate with the cast unwrapped
+                        if (!(lowBoundUnwrappedComparison.right() instanceof Constant(Type _, Object lowBoundValue))
+                                || !(highBoundUnwrappedComparison.right() instanceof Constant(Type _, Object highBoundValue))) {
+                            return expression;
```
`and(low, high)` is now the fallback when the two comparisons do not tighten into an inclusive range (wrong direction, or no adjacent value for a strict bound).

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -179,6 +201,119 @@ public Expression rewriteCall(Call node, Void context, ExpressionTreeRewriter<Vo
+            }
+
+            // Each surviving bound must be a comparison of the source against a constant; otherwise leave the
+            // BETWEEN untouched.
+            Optional<ComparisonBound> lowBound = comparisonBound(low, source);
+            Optional<ComparisonBound> highBound = comparisonBound(high, source);
+            if (lowBound.isEmpty() || highBound.isEmpty()) {
+                return Optional.empty();
+            }
+
+            // Rebuild an inclusive BETWEEN when the two comparisons form a lower/upper pair whose strict ends have
+            // an inclusive neighbor; otherwise keep them as a conjunction.
+            Optional<Object> inclusiveLow = inclusiveLowBound(sourceType, lowBound.get());
+            Optional<Object> inclusiveHigh = inclusiveHighBound(sourceType, highBound.get());
+            if (inclusiveLow.isEmpty() || inclusiveHigh.isEmpty()) {
+                return Optional.of(and(low, high));
+            }
```
Not making this change. The fallback duplicates the cast source, but that matches the existing invariant of this rule rather than departing from it: the single-comparison unwraps of a `CAST(x AS date)` predicate already do the same. `EQUAL` produces `and(x >= a, x < b)`, `NOT_EQUAL` an `or` of two, and `IDENTICAL` an `and` of three, none guarded by determinism (see `unwrapTimestampToDateCast`). `unwrapCast` treats the cast source as safe to duplicate, and a non-deterministic predicate would not be pushed into a scan anyway. Guarding only the BETWEEN fallback would be inconsistent with the far more common single-comparison paths. If source duplication is ever considered a real problem it should be addressed rule-wide, not in this PR.

raunaqmorarka replied:
Correction to my previous reply. Verifying with a plan test showed the concern is real, though not in the BETWEEN fallback: the fallback rebuilds a non-trivial source through `IrExpressions.between`, which re-binds it in a `Let`, so it is evaluated once. But the pre-existing single-comparison path is genuinely broken: `CAST(nd AS date) = DATE '...'` unwraps to `nd >= t1 AND nd < t2`, repeating the source. Fixed rule-wide in a separate commit by skipping the unwrap when the cast source is non-deterministic, with a regression test. That entry guard also covers the BETWEEN fallback.

martint replied:
I'd push back on the rule-wide determinism guard as the fix. Skipping the unwrap for a non-deterministic source isn't the invariant to reach for — binding the source once is, which is exactly what this PR already does for the BETWEEN value through `IrExpressions.between`/`Let`. The single-comparison paths should be fixed the same way, not by bolting a second, different mechanism onto the rule.

Binding once is strictly better than the skip:

- The guard only helps *non-deterministic* sources. A deterministic-but-expensive source — `date_trunc(unit, some_udf(x))` — is still duplicated and double-evaluated under it. A `Let` evaluates it once regardless.
- It also throws away the unwrap, and the out-of-range/range simplifications that come with it, for non-deterministic sources, for no gain.
- It costs nothing in pruning — which is the reason the skip looked acceptable ("wouldn't be pushed into a scan anyway"). `Reference`/`Constant` sources stay inline, so `DomainTranslator` still extracts a `Domain`; and `DomainTranslator.visitLet` inlines a bound source symbolically, so a `Let` around a non-trivial source is pruning-transparent. A non-trivial source was never prunable, so binding it loses nothing — it just makes it correct.

Net: bind the source once in the single-comparison paths, mirroring `SimplifyContinuousInValues` and the BETWEEN value here, and the separate `isDeterministic` guard becomes unnecessary — one mechanism, correct for deterministic-expensive and non-deterministic sources alike.

I put up #30286 doing exactly this for the timestamp→date paths (`EQUAL`/`NOT_EQUAL`/`IDENTICAL`), with tests. It's independent of the BETWEEN work here; once it lands, the guard here can come out and this fallback already aligns.
